### PR TITLE
Rotate stirrups with structure

### PR DESCRIPTION
**Pull request by SurajDadral**
Add functionality to rotate Stirrups with Structure
Code snippet:
structureAngle = math.degrees(structure.Placement.Rotation.Angle)
if structureAngle in {0.0, 180.0}:
    stirrupAngle = structureAngle
else:
    stirrupAngle = structureAngle - 90
rebar.Placement.Rotation.__setattr__("Angle", math.radians(stirrupAngle))

Illustration:
Before this code:
if column makes angle, say 45 degree, with x-axis and top view of column looks like:

TopViewBefore1

And we create stirrups by selecting top face of column, then top view of column is:

image

**After using this code:**

image

### CODE REVIEW

File: Stirrup.py [hunk 1]
```diff
@@ -237,6 +237,15 @@ def makeStirrup(l_cover, r_cover, t_cover, b_cover, f_cover, bentAngle, bentFact
...
     rebar.Direction = FaceNormal.negative()
     rebar.Rounding = rounding
+
+    # Rotate Stirrups with Structure
+    structureAngle = math.degrees(structure.Placement.Rotation.Angle)
+    if structureAngle in {0.0, 180.0}:
+        stirrupAngle = structureAngle
+    else:
+        stirrupAngle = structureAngle - 90
+    rebar.Placement.Rotation.__setattr__("Angle", math.radians(stirrupAngle))
```
Comment (amrit3701) [inline Stirrup.py]: Here don't need to use `__setter__` method. Use `rebar.Placement.Rotation.Angle = math.radians(stirrupAngle)`.

File: Stirrup.py [hunk 2]
```diff
@@ -237,6 +237,15 @@ def makeStirrup(l_cover, r_cover, t_cover, b_cover, f_cover, bentAngle, bentFact
             int((size - diameter) / amount_spacing_value), f_cover)
     rebar.Direction = FaceNormal.negative()
     rebar.Rounding = rounding
+
+    # Rotate Stirrups with Structure
+    structureAngle = math.degrees(structure.Placement.Rotation.Angle)
+    if structureAngle in {0.0, 180.0}:
```
Comment (amrit3701) [inline Stirrup.py]: Here `tuple` is better option. `(0.0, 180.0,)`

File: Stirrup.py [hunk 3]
```diff
@@ -321,6 +330,15 @@ def editStirrup(Rebar, l_cover, r_cover, t_cover, b_cover, f_cover, bentAngle, b
     Rebar.TopCover = t_cover
     Rebar.BottomCover = b_cover
     Rebar.TrueSpacing = amount_spacing_value
+
+    # Rotate Stirrups with Structure
+    structureAngle = math.degrees(structure.Placement.Rotation.Angle)
+    if structureAngle in {0.0, 180.0}:
```
Comment (amrit3701) [inline Stirrup.py]: Same comment.

File: Stirrup.py [hunk 4]
```diff
@@ -321,6 +330,15 @@ def editStirrup(Rebar, l_cover, r_cover, t_cover, b_cover, f_cover, bentAngle, b
...
+        stirrupAngle = structureAngle
+    else:
+        stirrupAngle = structureAngle - 90
+    Rebar.Placement.Rotation.__setattr__("Angle", math.radians(stirrupAngle))
```
Comment (amrit3701) [inline Stirrup.py]: Here also.

**Comment by SurajDadral** [comment]
Done all above suggested changes in commit https://github.com/amrit3701/FreeCAD-Reinforcement/pull/14/commits/02f5164a3cbe50a478f30a39ee0b320122991682

**Comment by SurajDadral** [comment]
> Also, write more description about this PR. Maybe use images.

Done

**Comment by SurajDadral** [comment]
> Its works fine only when we are rotating structure around z-axis.

This is because in most of the cases, columns rotates only about z-axis. In rare cases, columns are made inclined.

**Comment by SurajDadral** [comment]
> Can you explain the logic behind this?

I have proceeded as follows to develop logic:
- Firstly, I added line ```line.Placement.Rotation.Angle = structure.Placement.Rotation.Angle``` line in function makeStirrup() and started testing the code.
- I works fine if length and width of structure is same.
- If length and width of structure are not equal then it gives output as follows:
    - If 0<=structureAngle<=30 then code runs fine.
    - If 30<structureAngle<150 or 210<structureAngle<330 then it places stirrups perpendicular to structure as here (top view):

    image

To overcome this issue, I decided to subtract 90 from stirrup angle when 30<structureAngle<150 or 210<structureAngle<330

So the code becomes:
if 30 < structureAngle < 150 or 210 < structureAngle <= 330:
    wireAngle = structureAngle - 90
else:
    wireAngle = structureAngle

**Comment by SurajDadral** [comment]
> I think we can do it by other way also. Just rotate rebar at structure angle. For eg.
> `Rebar.Placement.Rotation = structure.Placement.Rotation`

It also produces same problem as I discussed above.

**Pull request closed without merging**